query-engine/query-engine-c-abi/src/engine.rs

```diff
@@ -0,0 +1,674 @@
+use crate::{
+    error::ApiError,
+    logger::Logger,
+    migrations::{
+        execute_migration_script, list_migration_dir, list_migrations, record_migration_started, MigrationDirectory,
+    },
+};
+use once_cell::sync::Lazy;
+use query_core::{
+    protocol::EngineProtocol,
+    schema::{self},
+    telemetry, TransactionOptions, TxId,
+};
+use request_handlers::{load_executor, RequestBody, RequestHandler};
+use serde_json::json;
+use std::{
+    env,
+    ffi::{c_char, c_int, CStr, CString},
+    path::{Path, PathBuf},
+    ptr::null_mut,
+    sync::Arc,
+};
+use tokio::{
+    runtime::{self, Runtime},
+    sync::RwLock,
+};
+use tracing::{field, instrument::WithSubscriber, level_filters::LevelFilter, Instrument, Span};
+use user_facing_errors::Error;
+
+use query_engine_common::engine::{
+    stringify_env_values, ConnectedEngine, ConnectedEngineNative, EngineBuilder, EngineBuilderNative, Inner,
+};
+use request_handlers::ConnectorKind;
+// use tracing_subscriber::filter::LevelFilter;
+
+// The query engine code is async by nature, however the C API does not function with async functions
+// This runtime is here to allow the C API to block_on it and return the responses in a sync manner
+static RUNTIME: Lazy<Runtime> = Lazy::new(|| runtime::Builder::new_multi_thread().enable_all().build().unwrap());
+
+// C-like return codes
+#[no_mangle]
+pub static PRISMA_OK: i32 = 0;
+#[no_mangle]
+pub static PRISMA_UNKNOWN_ERROR: i32 = 1;
+#[no_mangle]
+pub static PRISMA_MISSING_POINTER: i32 = 2;
+
+/// This struct holds an instance of the prisma query engine
+/// You can instanciate as many as you want
+pub struct QueryEngine {
+    // This String pointer will be pointed to by passed error pointer and will allow calling code to have a bit more information when something goes wrong
+    error_string: RwLock<CString>,
```

Comment:
I placed it there to prevent de-allocation, since on the API calls the calling code only passes a pointer. It's the same pattern that SQLite uses. Internally reserving memory for a global error str, which is needed to interact with C code

Reply:
There's something I don't understand with this pattern. Say you do the following:

- call `query` and get back an allocated string A
- call `query` again and get back another allocated string B

Now you have allocated two strings in Rust and C side has two references to these strings. However, on the second call to `query`, you have overwritten the content of `QueryEngine::error_string`. So what happens to allocated string A ? Won't it be pointing to freed memory since the content of `QueryEngine::error_string` will be replaced by allocated string B ?

Reply:
Both error pointers would point to the same error string. This is however the only way to share the memory between the contexts on the C world. It is the same pattern sqlite uses to return user-readable errors. 

https://www.sqlite.org/c3ref/errcode.html

![Oscar Franco Screen 001222](https://github.com/prisma/prisma-engines/assets/1634213/99188826-d266-4e25-8aa4-40b15d0b9303)

Otherwise the calling context next to allocate a sizeable enough buffer and needs to take care of cleaning the memory etc.

Reply:
@ospfranco 

> Both error pointers would point to the same error string.

Actually no, you are overwriting the whole `CString` struct:

```
                let mut error_string = query_engine.error_string.write().await;
                *error_string = CString::new(err.to_string()).unwrap();
```

So yes, the address of the value behind `RwLock` stays the same (the value being a `CString` struct itself) but the address of the raw string (which the `CString` smart pointer points to) will be a different allocation with a different address each time. The previous pointer will become dangling as soon as the new `CString` value is written (and, consequently, the old value is dropped).

Reply:
Yeah, the old value is dropped. I meant the pointer on the calling C context should always have a reference to the string in the struct, right? Or did I do it wrong?

Reply:
I believe Alexey's right here. If you look [at the documentation](https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.as_ptr) for `as_ptr()`, it says:

**WARNING**

The returned pointer is read-only; writing to it (including passing it to C code that writes to it) causes undefined behavior.

It is your responsibility to make sure that the underlying memory is not freed too early. For example, the following code will cause undefined behavior when ptr is used inside the unsafe block:

```rust
use std::ffi::CString;

// Do not do this:
let ptr = CString::new("Hello").expect("CString::new failed").as_ptr();
unsafe {
    // `ptr` is dangling
    *ptr;
}
```

This happens because the pointer returned by as_ptr does not carry any lifetime information and the CString is deallocated immediately after the CString::new("Hello").expect("CString::new failed").as_ptr() expression is evaluated. To fix the problem, bind the CString to a local variable:

```rust
use std::ffi::CString;

let hello = CString::new("Hello").expect("CString::new failed");
let ptr = hello.as_ptr();
unsafe {
    // `ptr` is valid because `hello` is in scope
    *ptr;
}
```
This way, the lifetime of the CString in hello encompasses the lifetime of ptr and the unsafe block.

Reply:
Disclaimer: it's the first time I'm dealing with C-Rust FFI code, so bear with me as I propose something that could be stupid:

I still don't understand why we need a `RwLock<CString>`, given that the entire API has to be sync. My suggestion is the following:

1. Store the last error as a `CString` in the QueryEngine
2. Mutate the `QueryEngine` struct in the C API and set `query_engine.last_error`. eg:
```rust
pub unsafe extern "C" fn prisma_apply_pending_migrations(
    qe: *mut QueryEngine,
    migration_folder_path: *const c_char,
) -> c_int {
    let mut query_engine: Box<QueryEngine> = Box::from_raw(qe);
    let result = RUNTIME.block_on(async { query_engine.apply_migrations(migration_folder_path).await });
    match result {
        Ok(_) => {
            query_engine.last_error = None;
            std::mem::forget(query_engine);
            PRISMA_OK
        }
        Err(err) => {
            query_engine.last_error = Some(CString::new(err.to_string()).unwrap());
            std::mem::forget(query_engine);
            PRISMA_UNKNOWN_ERROR
        }
    }
}
```
3. Add a new API function called `prisma_last_error(qe: *mut QueryEngine): *const c_char`, similar to what SQLite exposes:
```rust
#[no_mangle]
pub unsafe extern "C" fn prisma_last_error(qe: *mut QueryEngine) -> *const c_char {
    let query_engine: Box<QueryEngine> = Box::from_raw(qe);

    match query_engine.last_error {
        Some(str) => str.as_ptr() as *const c_char,
        None => std::ptr::null(),
    }
}
```
4. Remove all `error_string_ptr: *mut *mut c_char` C-API parameters

With this pattern, we don't have to constantly juggle between CString and pointers, potentially dropping stuff that we shouldn't. The C side has to explicitly call `prisma_last_error` to get an allocated string of the last error. The contract, similar to SQLite, is that it should never mutate it. `last_error`'s memory is always freed rust side. It's C side that has to be careful not to re-use an error string pointer after calling another method from the library. wdyt?

Reply:
After reviewing this bit again with @aqrln, we concluded two things:

- Even though the ABI is sync, there's no guarantee that the C side won't call the ABI from concurrent threads. Therefore, we can't just get rid of the `RwLock` like I suggested.
- However, with the current design (not the one I suggested), the `RwLock` serves no purpose in a concurrent thread scenario, because we can have multiple concurrent _readers_ and two threads might try to read the same error at the same time. Given that Rust is currently responsible for freeing the memory, I'm not sure whether it would end with C having a null ptr or a ptr pointing to the wrong error message for one of the concurrent call. And if we used a `Mutex` instead, we'd have to pass the Mutex guard over to C to keep it thread-safe. Not even sure that's possible, and I don't think we want that either.

We think the _only_ way to remain thread-safe with the ABI, even though _your_ usage from RN might not need that atm, is to:
1. Pass a `error_string_ptr: *mut *mut c_char` parameter to all functions of the ABI
2. Allocate the error string Rust-side and use `.into_raw()` to explicitly give ownership to C-side
3. Let C-side free that string when it doesn't need it anymore

This also solves the problem of `prisma_create` when you couldn't allocate an error string and had to `println!` instead because the `QueryEngine` struct wasn't created yet.

Reply:
Done!